components/resource_metering/src/agent.rs

```diff
@@ -0,0 +1,193 @@
+    fn default() -> Config {
+        Config {
+            enabled: false,
+            agent_address: "".to_string(),
+            precision_seconds: 1,
+            max_resource_groups: 5000,
+        }
+    }
+}
+
+impl Config {
+    pub fn validate(&self) -> std::result::Result<(), Box<dyn std::error::Error>> {
+        if !self.agent_address.is_empty() {
+            tikv_util::config::check_addr(&self.agent_address)?;
+        }
+
+        if self.precision_seconds < MIN_PRECISION_SECONDS
+            || self.precision_seconds > MAX_PRECISION_SECONDS
+        {
+            return Err(format!(
+                "precision seconds must between {} and {}",
+                MIN_PRECISION_SECONDS, MAX_PRECISION_SECONDS
+            )
+            .into());
+        }
+
+        if self.max_resource_groups > MAX_MAX_RESOURCE_GROUPS {
+            return Err(format!(
+                "max resource groups must between {} and {}",
+                0, MAX_MAX_RESOURCE_GROUPS
+            )
+            .into());
+        }
+
+        Ok(())
+    }
+}
+
+pub struct ConfigManager {
+    current_config: Config,
+    scheduler: Scheduler<Task>,
+}
+
+impl ConfigManager {
+    pub fn new(current_config: Config, scheduler: Scheduler<Task>) -> Self {
+        ConfigManager {
+            current_config,
+            scheduler,
+        }
+    }
+}
+
+impl configuration::ConfigManager for ConfigManager {
+    fn dispatch(
+        &mut self,
+        change: ConfigChange,
+    ) -> std::result::Result<(), Box<dyn std::error::Error>> {
+        self.current_config.update(change);
+        self.current_config.validate()?;
+
+        self.scheduler
+            .schedule(Task::ConfigChange(self.current_config.clone()))
+            .ok();
+
+        Ok(())
+    }
+}
+
+pub struct CpuRecordsCollector {
+    scheduler: Scheduler<Task>,
+}
+
+impl CpuRecordsCollector {
+    pub fn new(scheduler: Scheduler<Task>) -> Self {
+        Self { scheduler }
+    }
+}
+
+impl Collector for CpuRecordsCollector {
+    fn collect(&self, records: Arc<CpuRecords>) {
+        self.scheduler.schedule(Task::CpuRecords(records)).ok();
+    }
+}
+
+pub enum Task {
+    ConfigChange(Config),
+    CpuRecords(Arc<CpuRecords>),
+}
+
+impl Display for Task {
+    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
+        match self {
+            Task::ConfigChange(_) => {
+                write!(f, "ConfigChange")?;
+            }
+            Task::CpuRecords(_) => {
+                write!(f, "CpuRecords")?;
+            }
+        }
+
+        Ok(())
+    }
+}
+
+pub struct ResourceMeteringAgent {
+    config: Config,
+
+    env: Arc<Environment>,
+    security_mgr: Arc<SecurityManager>,
+
+    // TODO: mock client for testing
+    client: Option<ResourceUsageAgentClient>,
+    cpu_records_collector: Option<CollectorHandle>,
+}
+
+impl ResourceMeteringAgent {
+    pub fn new(
+        config: Config,
+        env: Arc<Environment>,
+        security_mgr: Arc<SecurityManager>,
```

Comment:
Do we need some separated security config? Since as a collect agent it only need to accept connections from this TiKV and it doesn't need intra-cluster connections.